storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java

```diff
@@ -440,9 +442,23 @@ public void recoverRunningTopology(final LocalAssignment currentAssignment, fina
      *
      * @param assignment the assignment the resources are for
      * @param port       the port the topology is running on
+     * @param downloadingBlobs any existing downloading blob futures
      * @throws IOException on any error
      */
-    public void releaseSlotFor(LocalAssignment assignment, int port) throws IOException {
+    public void releaseSlotFor(LocalAssignment assignment, int port,
+                               Collection<Future<Void>> downloadingBlobs) throws IOException {
+
+        // Make sure any downloading blobs in the background are stopped.
+        // This prevents a race condition where we could be adding references to a
+        // delayed downloading blob after the slot gets released, causing orphaned blobs.
+        for (Future future : downloadingBlobs) {
+            if (!future.isDone()) {
+                LOG.info("Canceling download of {}", future);
+                future.cancel(true);
+            }
+        }
+        downloadingBlobs.clear();
```

Comment:
Can this be a instance method in DynamicState?
This instance method will also need to be called from drainAllChangingBlobs() just before a new DynamicState is returned.

Reply:
I agree drainAllChangingBlobs() should also clean this up.  I will look into this.  Good find.